universal-application-tool-0.0.1/app/controllers/admin/AdminProgramController.java

```diff
@@ -0,0 +1,51 @@
+package controllers.admin;
+
+import javax.inject.Inject;
+import play.data.DynamicForm;
+import play.data.FormFactory;
+import play.mvc.Controller;
+import play.mvc.Http.Request;
+import play.mvc.Result;
+import services.program.ProgramService;
+import views.admin.ProgramList;
+import views.admin.ProgramNewOne;
+
+/**
+ * This controller contains an action to handle HTTP requests to the application's Admin "Create
+ * application" page.
+ */
+public class AdminProgramController extends Controller {
+
+  private final ProgramService service;
+  private final ProgramList listView;
+  private final ProgramNewOne newOneView;
+  private final FormFactory formFactory;
+
+  @Inject
+  public AdminProgramController(
+      ProgramService service,
+      ProgramList listView,
+      ProgramNewOne newOneView,
+      FormFactory formFactory) {
+    this.service = service;
+    this.listView = listView;
+    this.newOneView = newOneView;
+    this.formFactory = formFactory;
+  }
+
+  public Result list() {
+    return ok(listView.render(this.service.listProgramDefinitions()));
+  }
+
+  public Result newOne(Request request) {
+    return ok(newOneView.render(request));
+  }
+
+  public Result create(Request request) {
+    DynamicForm requestData = this.formFactory.form().bindFromRequest(request);
+    String name = requestData.get("name");
+    String description = requestData.get("description");
+    service.createProgramDefinition(name, description);
+    return list();
```

Comment:
question (blocking): I'm wondering if instead of returning `list()` we may want to return a different `Result`, such as a redirect like:

```java
return redirect("/admin/programs");
```

[Play docs](https://www.playframework.com/documentation/2.8.x/JavaActions#Redirects-are-simple-results-too) on redirect Results.

I also just found all of these [`created` methods](https://github.com/playframework/playframework/blob/master/core/play/src/main/java/play/mvc/Results.java#L776) on `Results`. The 201 "created" response type seems like the right thing semantically, but it doesn't looks like you can pass it a URL string to redirect to, so it might not be right in this case where we just want to send the user back to the list page.

Not sure if it matters, but thought I would flag this in case there is some "proper" response type for create/POSTs.

Reply:
Definitely redirect rather than directly calling the list method. Recommend using a 302 redirect: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/302

Reply:
using `Results.found` for 302.

Reply:
The tests had to change for the redirect. They now test that the response is 302 FOUND, and then makes another request and checks the list of programs to see if the new programs show up.